### PR TITLE
SRE-33 haskell overhaul

### PR DESCRIPTION
**Pull request by craigem**
This is the overhaul of Haskell for `cardano-base` along the lines of `cardano-node` [PR #507](https://github.com/input-output-hk/cardano-node/pull/507/).

### CODE REVIEW

File: lib.nix [hunk 1]
```diff
@@ -1,18 +1,42 @@
...
+  haskellPackages = import ./nix/pkgs.nix {
+    inherit pkgs;
+    src = ./.;
+  };
   lib = pkgs.lib;
-in lib // { inherit iohkNix pkgs; inherit (iohkNix) nix-tools; }
+  niv = (import sources.niv {}).niv;
+  isCardanoBase = with lib; package:
+    (package.isHaskell or false) &&
+      (hasPrefix "cardano-base" package.identifier.name);
```
Comment (disassembler) [inline lib.nix]: need to fix the prefixes here. none of these packages start with cardano-base

File: scripts/buildkite/cache-s3.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{ stdenv, fetchurl, zlib, gmp }:
```
Comment (disassembler) [inline scripts/buildkite/cache-s3.nix]: why are you adding this file?

File: stack.yaml [hunk 3]
```diff
@@ -1,17 +1,117 @@
-resolver: https://raw.githubusercontent.com/input-output-hk/cardano-prelude/cb54482c7e3d45f6c1aac490dc937ed835de0333/snapshot.yaml
+resolver: https://raw.githubusercontent.com/input-output-hk/cardano-prelude/00487726c4bc21b4744e59d913334ebfeac7d68e/snapshot.yaml
```
Comment (disassembler) [inline stack.yaml]: this absolutely looks wrong

File: scripts/buildkite/rebuild.hs [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {-# LANGUAGE LambdaCase          #-}
```
Comment (disassembler) [inline scripts/buildkite/rebuild.hs]: is this file being changed intentionally?

Comment (craigem) [reply scripts/buildkite/rebuild.hs]: reverted this back to the version from master with #9025568

File: scripts/buildkite/cache-s3.nix (hunk 2 above)
Comment (craigem) [reply scripts/buildkite/cache-s3.nix]: It's called from default.nix.

Removed the file and it's reference form `default.nix` in dba088f

File: stack.yaml (hunk 3 above)
Comment (craigem) [reply stack.yaml]: Reverted to the resolver from master with commit 860e263

File: lib.nix (hunk 1 above)
Comment (craigem) [reply lib.nix]: Prefix removed in commit 729da10

**Comment by disassembler** [comment]
bors ping

**Pull request merged**